bundles/org.openhab.binding.energidataservice/src/main/java/org/openhab/binding/energidataservice/internal/action/EnergiDataServiceActions.java

```diff
@@ -0,0 +1,381 @@
+                throw new IllegalArgumentException(String.format("'%s' has no corresponding value. Accepted values: %s",
+                        name, Arrays.asList(values())));
+            }
+            return myEnum;
+        }
+    }
+
+    @RuleAction(label = "@text/action.get-prices.label", description = "@text/action.get-prices.description")
+    public @ActionOutput(name = "prices", type = "java.util.Map<java.time.Instant, java.math.BigDecimal>") Map<Instant, BigDecimal> getPrices() {
+        return getPrices(Arrays.stream(PriceElement.values()).collect(Collectors.toSet()));
+    }
+
+    @RuleAction(label = "@text/action.get-prices.label", description = "@text/action.get-prices.description")
+    public @ActionOutput(name = "prices", type = "java.util.Map<java.time.Instant, java.math.BigDecimal>") Map<Instant, BigDecimal> getPrices(
+            @ActionInput(name = "priceElements", label = "@text/action.get-prices.priceElements.label", description = "@text/action.get-prices.priceElements.description") @Nullable String priceElements) {
+        if (priceElements == null) {
+            logger.warn("Argument 'priceElements' is null");
+            return Map.of();
+        }
+
+        Set<PriceElement> priceElementsSet;
+        try {
+            priceElementsSet = new HashSet<PriceElement>(
+                    Arrays.stream(priceElements.split(",")).map(PriceElement::fromString).toList());
+        } catch (IllegalArgumentException e) {
+            logger.warn("{}", e.getMessage());
+            return Map.of();
+        }
+
+        return getPrices(priceElementsSet);
+    }
+
+    @RuleAction(label = "@text/action.calculate-price.label", description = "@text/action.calculate-price.description")
+    public @ActionOutput(name = "price", type = "java.math.BigDecimal") BigDecimal calculatePrice(
+            @ActionInput(name = "start", type = "java.time.Instant") Instant start,
+            @ActionInput(name = "end", type = "java.time.Instant") Instant end,
+            @ActionInput(name = "power", type = "QuantityType<Power>") QuantityType<Power> power) {
+        PriceCalculator priceCalculator = new PriceCalculator(getPrices());
+
+        try {
+            return priceCalculator.calculatePrice(start, end, power);
+        } catch (MissingPriceException e) {
+            logger.warn("{}", e.getMessage());
+            return BigDecimal.ZERO;
+        }
+    }
+
+    @RuleAction(label = "@text/action.calculate-cheapest-period.label", description = "@text/action.calculate-cheapest-period.description")
+    public @ActionOutput(name = "result", type = "java.util.Map<String, Object>") Map<String, Object> calculateCheapestPeriod(
+            @ActionInput(name = "earliestStart", type = "java.time.Instant") Instant earliestStart,
+            @ActionInput(name = "latestEnd", type = "java.time.Instant") Instant latestEnd,
+            @ActionInput(name = "duration", type = "java.time.Duration") Duration duration) {
```

Comment:
@florian-h05 - perhaps you can advise, (how) is it possible to use these method overloads from JavaScript? I initially tested with DSL rules, but now trying JS:

```javascript
edsActions = actions.get("energidataservice", "energidataservice:service:energidataservice");
var result = edsActions.calculateCheapestPeriod(time.Instant.now(), time.Instant.now().plusSeconds(12*60*60), time.Duration.ofMinutes(90));
```

Result:
```
2023-05-09 22:51:45.308 [ERROR] [enhab.automation.script.file.test.js] - Failed to execute rule Test-EDS-d01ebd58-3ae8-41a8-80cc-769811a64f73: TypeError: invokeMember (calculateCheapestPeriod) on org.openhab.binding.energidataservice.internal.action.EnergiDataServiceActions@7e4decc9 failed due to: no applicable overload found (overloads: [Method[public java.util.Map org.openhab.binding.energidataservice.internal.action.EnergiDataServiceActions.calculateCheapestPeriod(java.time.Instant,java.time.Instant,java.time.Duration,java.util.List,org.openhab.core.library.types.QuantityType)], Method[public java.util.Map org.openhab.binding.energidataservice.internal.action.EnergiDataServiceActions.calculateCheapestPeriod(java.time.Instant,java.time.Instant,java.util.List,java.util.List)], Method[public java.util.Map org.openhab.binding.energidataservice.internal.action.EnergiDataServiceActions.calculateCheapestPeriod(java.time.Instant,java.time.Instant,java.time.Duration,org.openhab.core.library.types.QuantityType)], Method[public java.util.Map org.openhab.binding.energidataservice.internal.action.EnergiDataServiceActions.calculateCheapestPeriod(java.time.Instant,java.time.Instant,java.time.Duration)]], arguments: [DynamicObject<JSOrdinary>@40d066e5 (DefaultLayout), DynamicObject<JSOrdinary>@312634d9 (DefaultLayout), DynamicObject<JSOrdinary>@7cd571ea (DefaultLayout)]): TypeError: invokeMember (calculateCheapestPeriod) on org.openhab.binding.energidataservice.internal.action.EnergiDataServiceActions@7e4decc9 failed due to: no applicable overload found (overloads: [Method[public java.util.Map org.openhab.binding.energidataservice.internal.action.EnergiDataServiceActions.calculateCheapestPeriod(java.time.Instant,java.time.Instant,java.time.Duration,java.util.List,org.openhab.core.library.types.QuantityType)], Method[public java.util.Map org.openhab.binding.energidataservice.internal.action.EnergiDataServiceActions.calculateCheapestPeriod(java.time.Instant,java.time.Instant,java.util.List,java.util.List)], Method[public java.util.Map org.openhab.binding.energidataservice.internal.action.EnergiDataServiceActions.calculateCheapestPeriod(java.time.Instant,java.time.Instant,java.time.Duration,org.openhab.core.library.types.QuantityType)], Method[public java.util.Map org.openhab.binding.energidataservice.internal.action.EnergiDataServiceActions.calculateCheapestPeriod(java.time.Instant,java.time.Instant,java.time.Duration)]], arguments: [DynamicObject<JSOrdinary>@40d066e5 (DefaultLayout), DynamicObject<JSOrdinary>@312634d9 (DefaultLayout), DynamicObject<JSOrdinary>@7cd571ea (DefaultLayout)])
	at <program> (test.js:10)
	at _run (@openhab-globals.js:2)
	at execute (@openhab-globals.js:2)
	at execute (@openhab-globals.js:2)
```

I would like to provide all examples in JavaScript as well in the README, but struggling a bit.

Reply:
Normally these method overloads work, however it seems that there is an argument type mismatch because of missing "translations" from JS to Java types.
I've just checked the JS Scripting add-on code, we only have "translations" in place for ZonedDateTime and Duration, but not for Instant. Therefore the add-on needs adjustment.

I can have a look into this, should not be too much work or I can give you advice how to do that yourself

Reply:
If you mean add "translation" for `Instant` in the JS Scripting add-on, that would be awesome as it would be a generic fix for all bindings. If you are willing to do this for 4.0, I would highly appreciate that and be grateful. 🙂

I'm in no rush, just wanted to improve the docs, and eventually my plan is to migrate all my rules to JS anyway, so I might as well get started with the rules using the actions of this binding.

Reply:
I will add this for 4.0, I already have a few code improvements lying around so I’m going to open a PR soon. I’m currently working on the evcc binding, but when I’m finished I will take care of JS Scripting.

I have already migrated most rules to JS (seems reasonable since I am the one doing most work on JS Scripting), IMO it is pretty well developed now and I don’t see any larger changes on the horizon.

Reply:
@jlaur Do you think that transactions for LocalDate(Time) are also needed anywhere?

Reply:
@jlaur 
PR is now open, see #14984.
Can you please compile and test it?

Reply:
It works - thanks a lot!

Reply:
Great!